### DISCUSSION TITLE
When will 3.2.0 be released?

### DISCUSSION THREAD

**Discussion by msqr1** [emscripten-core/emscripten#21506] [General]
I know Emscripten is using semantic versioning, but the patch version is only for bug fixes, we've been adding new features like --use-ports, wouldn't that bump the minor counter? I feel like the patch counter is getting out of hand, it's at 55 now.

**Comment by tlively** [comment]
Emscripten does not use semantic versioning, even though our version numbers make it look like we do. We usually bump the minor counter when something suitably "big" changes.

**Comment by msqr1** [comment]
What versioning does it use? What count as a big change, and with this, will the major version ever get bumped?
@tlively

**Reply by sbc100** [reply]
Our versioning policy has changed over time, but its safe to say that its never really been semantic versioning.   I guess you could say its closer the linux kernel versioning scheme which is to say that its mostly and the whim of the developers. 

One reason for this is that almost every release of emscripten includes changes that could (potentially) break somebody, even though we do try to minimize breakages.  For example if you build with `-Wall -Werror` as many users do, then every time we add a new warning you will get broken.

That last time we discussed bumping the major or minor version @tlively actually suggested that perhasps we don't do this, so that it becomes easier to guess how far apart two versions are from each other (both in time and in version space).   If we continue with 3.1.XXX it then its easy to see the 3.1.50 and 3.1.70 are quite far about (perhaps about a year since we release every couple of weeks).   So you don't need to lookup the release to get a ball pack for how far from latest a given release is in time.   This is one of nice features of chrome's and firefox's version scheme which is single number.  If the current version of chrome 120 .. so its easy to know the chrome 70 is extremely old at this point in time.   Of course that begs the question, why not just switch to single number.. and indeed that might be something we should do.

For more discussion on this topic see: #15917

**Reply by kleisauke** [reply]
> 3.1.XXX

FWIW, hopefully Emscripten would not use 3-digit minor/patch version component, because that could cause issues on downstream projects. For example, with projects that are doing:
```c
if __EMSCRIPTEN_major__*10000 + __EMSCRIPTEN_minor__*100 + __EMSCRIPTEN_tiny__ >= 30156
```

**Reply by sbc100** [reply]
Right, but because we never go back and patch any existing releases (we only ever fix forward, so to speak) if we applied the the chome scheme we would indeed end up with single number because the other two would always be zero.

**Reply by msqr1** [reply]
Why not just switch to single number right now, if the other 2 numbers don't have any use?

**Reply by sbc100** [reply]
That is a good question, and one I think we should consider.

The discussion in https://github.com/emscripten-core/emscripten/issues/15917 has not concluded so I think we should think a little more on it.

**Reply by msqr1** [reply]
For a single version number though, what if a really big change land, we bump it by 2,3? And changing the versioning system like this, we have to define a new macro like __EMSCRIPTEN_VERSION__, and that would make this in itself a breaking change for people testing downstream.

**Reply by sbc100** [reply]
I'm not sure how we would define "really big change".  We don't currently have any such definition.   I guess we did this one in the entire history of the project when we went from fastcomp to llvm backend.   But even that was supposed to be an un-observable change and should have been non-breaking for most folks.    So it was a really big internal change really.. not a really be external/breaking change.

I agree switching to single version would be tricky for reasons such as `EMSCRIPTEN_VERSION` macros.   My guess is that if we do that route we would just use `EMSCRIPTEN_VERSION_MAJOR` and fix the `MINIOR` and `PATCH`versions at zero.  However, we are still a long way from deciding to do that I think.

**Comment by sbc100** [comment]
> People know Chrome by that single number, but it doesn't actually use that. https://www.chromium.org/developers/version-numbers/



> > 3.1.XXX
> 
> FWIW, hopefully Emscripten would not use 3-digit minor/patch version component, because that could cause issues on downstream projects. For example, with projects that are doing:
> 
> ```c
> if __EMSCRIPTEN_major__*10000 + __EMSCRIPTEN_minor__*100 + __EMSCRIPTEN_tiny__ >= 30156
> ```

That seems like a weird assumption to make.. but sure we can try to avoid that.   

This code assumes the minor and tiny number can only go to 99 but the major version can go above 99..

**Reply by kleisauke** [reply]
Indeed, it's a somewhat unusual way to check for an specific Emscripten version at compile time. Some occurrences are:
https://github.com/rust-lang/libc/blob/a0f5b4b21391252fe38b2df9310dc65e37b07d9f/build.rs#L197
https://github.com/mosra/magnum/blob/b1419017650c83538d8fe4681de6f0bca524cf49/src/Magnum/Platform/EmscriptenApplication.h#L77
